File: alpha_main/scripts/dance/pid.py
class PID(object):
    def __init__(self, k_p=1.0, k_i=0.0, k_d=0.0):
        self.k_p = k_p
        self.k_i = k_i
        self.k_d = k_d

        self.e_i = 0.0
        self.e_d = 0.0

    def compute(self,err,dt):
        e_p = err
        e_i = self.e_i
        e_d = err - self.e_d

        self.e_i += err * dt
        self.e_d = err

        return self.k_p * e_p + self.k_i * e_i + self.k_d * e_d
    def reset(self):
        self.e_i = 0.0
        self.e_d = 0.0

File: alpha_main/src/pid.cpp
#include <cmath>
#include "alpha_main/pid.h"

PID::PID(float k_p, float k_i, float k_d):
	k_p(k_p),k_i(k_i),k_d(k_d)
{
	reset();
}

float PID::compute(float err, float dt){
	e_p = err;
	e_i += err * dt;
	e_d = (err - e_d);

	float res = k_p*e_p + k_i*e_i + k_d*e_d;
	e_d = err; // remember err for later
	return res;
}

void PID::reset(){
	e_p = e_i = e_d = 0.0;
}
